# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/PointerEventHandler.cpp
# Commit: ddd6c187d02d
# Full Hash: ddd6c187d02d1b2ece70f5c1501f508ad96f7604
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2025-05-15 22:58:00
# Regressor Bug: 1550462
# File Overlap Count: 1
# Description:
#   Bug 1960253 - Split the computation of target frame/content of events using coordinates from `EventHandler::HandleEventUsingCoordinates` r=smaug
#   
#   We'll need to compute the target `PresShell` before processing the pending
#   pointer capture to check whether the corresponding window needs
#   `pointerrawupdate` events (bug 1550462).
# ==============================================================================

diff -r cd42b1907239 -r ddd6c187d02d dom/events/PointerEventHandler.cpp
--- a/dom/events/PointerEventHandler.cpp	Wed May 14 20:47:11 2025 +0000
+++ b/dom/events/PointerEventHandler.cpp	Wed May 14 21:19:36 2025 +0000
@@ -564,8 +564,30 @@
 }
 
 /* static */
+Element* PointerEventHandler::GetPendingPointerCapturingElement(
+    uint32_t aPointerId) {
+  PointerCaptureInfo* pointerCaptureInfo = GetPointerCaptureInfo(aPointerId);
+  if (pointerCaptureInfo) {
+    return pointerCaptureInfo->mPendingElement;
+  }
+  return nullptr;
+}
+
+/* static */
 Element* PointerEventHandler::GetPointerCapturingElement(
-    WidgetGUIEvent* aEvent) {
+    const WidgetGUIEvent* aEvent) {
+  return GetPointerCapturingElementInternal(CapturingState::Override, aEvent);
+}
+
+/* static */
+Element* PointerEventHandler::GetPendingPointerCapturingElement(
+    const WidgetGUIEvent* aEvent) {
+  return GetPointerCapturingElementInternal(CapturingState::Pending, aEvent);
+}
+
+/* static */
+Element* PointerEventHandler::GetPointerCapturingElementInternal(
+    CapturingState aCapturingState, const WidgetGUIEvent* aEvent) {
   if ((aEvent->mClass != ePointerEventClass &&
        aEvent->mClass != eMouseEventClass) ||
       aEvent->mMessage == ePointerDown || aEvent->mMessage == eMouseDown) {
@@ -584,11 +606,13 @@
     return nullptr;
   }
 
-  WidgetMouseEvent* mouseEvent = aEvent->AsMouseEvent();
+  const WidgetMouseEvent* const mouseEvent = aEvent->AsMouseEvent();
   if (!mouseEvent) {
     return nullptr;
   }
-  return GetPointerCapturingElement(mouseEvent->pointerId);
+  return aCapturingState == CapturingState::Pending
+             ? GetPendingPointerCapturingElement(mouseEvent->pointerId)
+             : GetPointerCapturingElement(mouseEvent->pointerId);
 }
 
 /* static */